### DISCUSSION TITLE
Using actix-web Payload as ByteStream

### DISCUSSION THREAD

**Discussion by stefanhoelzl** [awslabs/aws-sdk-rust#458] [Q&A] [answered]
Hello, 
how is it possible to use an [`actix_web::web::Playload`](https://docs.rs/actix-web/latest/actix_web/web/struct.Payload.html) as  `input` parameter to [`aws_sdk_s3::client::fluent_builders::PutObject::body`](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/client/fluent_builders/struct.PutObject.html#method.body) accepting the type [`aws_sdk_s3::ByteStream`](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/struct.ByteStream.html)?

my [actix request handler](https://actix.rs/docs/handlers/) currently looks like this:
```rust
use actix_web::{post, web::Payload, HttpResponse, Responder};
use aws_config;
use aws_sdk_s3::{ByteStream, Client};

#[post("/")]
async fn create(body: Payload) -> impl Responder {
    let config = aws_config::load_from_env().await;
    let client = Client::new(&config);
    client
        .put_object()
        .bucket("my-bucket")
        .key("my-key")
        .body(ByteStream::from(body.to_vec())) // <- here I want to stream the body directly
        .send()
        .await
        .unwrap();
    HttpResponse::Ok()
}
```

[Here](https://actix.rs/actix-web/actix_web/web/struct.Payload.html#examples) is an example from the `Payload` documentation on how to use it as a stream, but I don't know how to use this for `aws_sdk_s3::ByteStream`

Thanks for your help!

**Comment by rcoh** [comment]
Hello! This is pattern we should probably add a better API for. You can do this via hyper::Body (note: Hyper Body does not use the Hyper client at all, just the body implementation):

```rust
let body: Payload = ....;
let body = hyper::Body::wrap_stream(body);
// convert hyper::Body into SdkBody
let byte_stream = ByteStream::new(body.into());
```

**Reply by stefanhoelzl** [reply]
Thanks a lot for your help!
I tried your suggestions but get the following error message:
```
error[E0277]: `Rc<RefCell<actix_http::h1::payload::Inner>>` cannot be sent between threads safely
   --> src/main.rs:21:43
    |
21  |     let stream = hyper::Body::wrap_stream(body);
    |                  ------------------------ ^^^^ `Rc<RefCell<actix_http::h1::payload::Inner>>` cannot be sent between threads safely
    |                  |
    |                  required by a bound introduced by this call
    |
    = help: within `actix_web::web::Payload`, the trait `Send` is not implemented for `Rc<RefCell<actix_http::h1::payload::Inner>>`
    = note: required because it appears within the type `actix_http::h1::payload::Payload`
    = note: required because it appears within the type `actix_web::dev::Payload`
    = note: required because it appears within the type `actix_web::web::Payload`
note: required by a bound in `hyper::Body::wrap_stream`
   --> /Users/stefan/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.17/src/body/body.rs:192:42
    |
192 |         S: Stream<Item = Result<O, E>> + Send + 'static,
    |                                          ^^^^ required by this bound in `hyper::Body::wrap_stream`

error[E0277]: `(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)` cannot be sent between threads safely
   --> src/main.rs:21:43
    |
21  |     let stream = hyper::Body::wrap_stream(body);
    |                  ------------------------ ^^^^ `(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)` cannot be sent between threads safely
    |                  |
    |                  required by a bound introduced by this call
    |
    = help: the trait `Send` is not implemented for `(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)`
    = note: required because of the requirements on the impl of `Send` for `Unique<(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)>`
    = note: required because it appears within the type `Box<(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)>`
    = note: required because it appears within the type `Pin<Box<(dyn futures_core::stream::Stream<Item = Result<actix_web::web::Bytes, PayloadError>> + 'static)>>`
    = note: required because it appears within the type `actix_web::dev::Payload`
    = note: required because it appears within the type `actix_web::web::Payload`
note: required by a bound in `hyper::Body::wrap_stream`
   --> /Users/stefan/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.17/src/body/body.rs:192:42
    |
192 |         S: Stream<Item = Result<O, E>> + Send + 'static,
    |                                          ^^^^ required by this bound in `hyper::Body::wrap_stream`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `backend` due to 2 previous errors
```
It fails when trying to wrap the `Payload` in a `hyper::Body`
```rust
let stream = hyper::Body::wrap_stream(body);
```

**Comment by rcoh** [comment]
Hmmm....I think this might be problematic. In general, the body definitely
_can_ be passed around to multiple threads (unless you're using a single
threaded runtime).

I'll need to investigate some more.

**Comment by rcoh** [comment, marked as answer]
I think you'll need something like SyncWrapper described in https://internals.rust-lang.org/t/what-shall-sync-mean-across-an-await/12020/2

The discussion is actually around this specific issue of a streaming HTTP body

**Comment by seabo** [comment]
Hello! I ran into the exact same issue as @stefanhoelzl (wanting to use an [actix_web::web::Playload](https://docs.rs/actix-web/latest/actix_web/web/struct.Payload.html) as an [aws_sdk_s3::types::ByteStream](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/types/struct.ByteStream.html)). 

Since the discussion above, it seems the [implementation of `hyper::Body::wrap_stream`](https://docs.rs/hyper/latest/src/hyper/body/body.rs.html#190-198) has incorporated the `SyncWrapper` idea @rcoh mentioned above. This means the `Sync` bound no longer exists on the `Stream` passed into `wrap_stream`. However, this approach still fails to compile because the `Send` bound remains, and [actix_web::web::Playload](https://docs.rs/actix-web/latest/actix_web/web/struct.Payload.html) is [`!Send`](https://docs.rs/actix-web/4.2.1/actix_web/web/struct.Payload.html#impl-Send-for-Payload).

I would appreciate any pointers on how to get this working.

**Comment by martijnarts** [comment]
I don't know if this problem is the same with actix-web, but I ran into a (similar) issue with Axum. I managed to solve it through an ownership-taking wrapper, as described in this blog post: https://blog.adamchalmers.com/streaming-proxy/
